### DISCUSSION TITLE
Implicit try block with finally/catch clause

### DISCUSSION THREAD

**Discussion by ActivistInvestor** [dotnet/csharplang#7837] [Language Ideas]
Not sure how to find out if this has been proposed already or not, but...

I would like to see support for implicit outer-most try/catch/finally blocks in the body of a method or function, wherein a finally and/or catch clause could appear without a try block, making the code that precedes either the catch or finally clause implicitly a try block, extending from the start of the method's body.

In other words:

```
public static bool SomeMethod()
{
   try
   {
      DoSomething();
      DoSomethingElse();
      return true;
   }
   catch(System.Exception ex)
   {
      HandleException(ex);
      return false;
   }
   finally
   {
      CleanUp();
   }
}

```
Could be expressed as:
```
public static bool SomeMethod()
{
   DoSomething();
   DoSomethingElse();
   return true;
   catch(System.Exception ex)
      HandleException(ex);
      return false;
   finally
      CleanUp();
}
```
The implied try{} block extends from the start of the
method, up to the first catch or finally 'statement'.

The try, catch and finally blocks are implicitly-
delimited, requiring no braces. The catch block
extends to either a finally 'statement' or the end 
of the method body. The finally block extends to 
the  end of the method body.

The purpose served is to minimize the complexity of 
retrofitting a try/catch/finally into existing code.

**Comment by siegfriedpammer** [comment]
Similar ideas:
#908
#968
#7127

And there are even more...

**Comment by colejohnson66** [comment]
If your goal is to remove a level of indentation, use seperate functions:

```cs
public static bool SomeMethod()
{
    try
    {
        SomeMethodCore();
        return true;
    }
    catch (Exception ex)
    {
        HandleException(ex);
        return false;
    }
    finally
    {
        CleanUp();
    }
}

private static void SomeMethodCore()
{
    DoSomething();
    DoSomethingElse();
}
```

> The try, catch and finally blocks are implicitly-delimited, requiring no braces.

Why? C# is not Python.

**Reply by ActivistInvestor** [reply]
The goal is the opposite of what you suggest:  Uncomplicate rather than complicate.  C# is not F# either, but one might not get that impression when looking at its evolution.

**Reply by tacosontitan** [reply]
> Uncomplicate rather than complicate.

I fail to see how this recommendation "complicates" the code. Extraction and inversion are very commonly applied techniques to improve the legibility of code, and explicit is frequently better than implicit.

**Reply by ActivistInvestor** [reply]
IMO, introducing additional levels of indirection if only for the purpose of avoiding an additional level of indentation resulting from the direct use of try and finally, is self-defeating because it also inherently complicates code, and is not justifiable. 

I extract code to a dedicated method when there is a just reason for doing it, such as when it needs to be reusable, or used from another context, etc.

**Reply by theunrepentantgeek** [reply]
You forgot one of the most important reasons to extract a method - to capture a specific operation with a well chosen name. This not only allows you to reason fully about the extracted method in isolation, it makes the original method easier to understand as well.

**Reply by ActivistInvestor** [reply]
If extraction helps keep a body of code from becoming too large, that's also a good reason, assuming that the extracted code is not being executed repetitively at a high frequency, which incurs a cost if it cannot be inlined.

If what is being proposed here is used in an unreasonably large body of code, the problem isn't the implicit try/catch/finally, the problem is the size of the body of code. 

I don't agree that extraction solely for canonical purposes is a good thing, especially if it doesn't get inlined. It may be okay for prototyping but not for a finished product.

Code comments and annotations have no impact on performance.

Furthermore, my original proposal was something that I envisioned only being used in relatively small bodies of code rather than large, complicated ones.

**Reply by tacosontitan** [reply]
> IMO, introducing additional levels of indirection if only for the purpose of avoiding an additional level of indentation resulting from the direct use of try and finally, is self-defeating because it also inherently complicates code, and is not justifiable.

If that's the only reason, then of course it's silly. Think of even just simple a case where you want to load some data from a file and write it to the console:

```cs
public sealed class Foo { public int Bar { get; set; } }

...
private readonly ISerializationProvider _serializationProvider;
public async Task Sample(CancellationToken cancellationToken)
{
    Foo? foo = await LoadFoo();
    if (foo is null)
    {
        Console.WriteLine("Unable to load foo.");
        return;
    }

    Console.WriteLine(foo.Bar);
}

private async Task<Foo?> LoadFoo()
{
    try
    {
        string content = await Task.Run(ReadAllText, cancellationToken);
        return _serializationProvider.Deserialize<T>(content);
    } catch (Exception e)
        Console.WriteLine(...);
        return null;
    }
    
    // Do I now need to move my local method to above the catch block if we support implicit `try-catch-finally`?
    // How much code is that going to bust up?
    string ReadAllText() =>
        File.ReadAllText("path to file");
}
```

Even in this scenario, you're only removing a few extra characters and I certainly don't like the idea of the `try` being implicit but the `catch` and `finally` being explicit. As such I'd want the `try` to be specified for consistency with explicit expectation which means you're only removing four or six braces and that, to me at least, is also silly.

**Reply by ActivistInvestor** [reply]
True, it only eliminates the try, and a code block.  My original proposal was not only to make the outer-most try/catch/finally implicit, but also to eliminate the need for the code block as well.

```
static bool CanOpenForWrite(string filePath)
{
   try
   {
      using new FileStream(filePath, FileMode.Open, FileAccess.ReadWrite);
      return true;
   }
   catch(System.Exception ex)
   {
      return false;
   }
}

Which could become:

static bool CanOpenForWrite(string filePath)
{
   using new FileStream(filePath, FileMode.Open, FileAccess.ReadWrite);
   return true;
   catch(System.Exception ex)
      return false;
}

```

I suppose one could have made the same/similar arguments against the implicit using() block as well.

**Reply by tacosontitan** [reply]
> I suppose one could have made the same/similar arguments against the implicit using() block as well.

I don't think so, to be honest. `using` blocks have been notorious for introducing wild levels of indentation with almost no need to do so thanks to nested `using` blocks and many repositories requiring braces for them prior to the introduction of the `using` statement.

You could extract each, but that would be back into the silly realm when you could just stack all but the last if you truly wanted to:

```cs
using (var a = new A())
using (var b = new B(a))
using (var c = new C(b))
using (var d = new D(c))
using (var e = new E(d))
{
    ...
}
```
The problem with this (in my experience) is that many people didn't like this, for the same reason they don't like nesting iterators in a similar fashion:

```cs
for (int x = 0; x < 100; x++)
for (int y = 0; y < 100; y++)
for (int z = 0; z < 100; z++)
{
    ...
}
```

As a result, we ended up with a way to reduce all of that nesting for `using` blocks in a meaningful way.

**Reply by CyrusNajmabadi** [reply]
> and many repositories requiring braces

That's an issue with those repositories.  It is a very intentional and purposeful part of our process that we do *not* design new langauge features to workaround whatever particular stylistic views a particular subset of users have. 

1. they are always free to get rid of that restriction.
2. that restriction is self inflicted.
3. there is no guarantee that whatever new feature we build will then be accepted by that group (or other groups) and not have just as many silly stylistic restrictions placed on it.

Trying to chase and appease people who do this is a specific *non-goal* for the language.

**Reply by ActivistInvestor** [reply]
I don't think this is so much about coding style, as it is about how a person that is new to the language, perhaps coming from Pascal, or another non-C/C++ -based language perceives things (FWIW, I came to C/C++, C# and Java from a LISP background 25 years ago, after some dabbling in Pascal, Haskell, and Forth).

One of the hardest things for a seasoned programmer to do is put themselves in the shoes of someone that is just starting out with the language, and in this case, has no prior exposure to C/C++, which is where some of the inconsistency comes from.

What inconsistency?

What does try, catch, and finally not have in common with these constructs:

do
while()
for()
foreach()
if()
etc....

**Reply by KennethHoff** [reply]
@ActivistInvestor the (ill-advised) option of an implicit single-expression scope? That's a fair argument actually. I don't see why it wouldn't be possible in this case as well.

**Reply by HaloFour** [reply]
> What does try, catch, and finally not have in common with these constructs

Parsing ambiguities.  As explicitly stated by [Bjarne Stroustrup](https://www.stroustrup.com/except89.pdf), even if the compiler could have rules which would make it unambiguous to parse, it would still be ambiguous and difficult to read.

> There seems to be no reason to allow these ambiguities even if there is a trivial and systematic way for a
parser to chose one interpretation over another.

**Reply by colejohnson66** [reply]
Specifically, the parsing ambiguity is due to the lack of parenthesis after the `try` and `do` keywords.

**Reply by Jeltz191** [reply]
I was somewhat playing devil's advocate in my suggestion, having a personal distaste for both implicit using and exception handling as done by C#. So I suggested abusing both ;-) To make a positive contribution:
I would rather elevate exceptions to the class level so as not to conflate the exception logic with the functional logic, there being an exception thread and a functional thread to our code, each with well understood responsibilities (by analogy, a bit how air traffic controllers manage aircraft movements around either aircraft or airport related problems). Exceptions would then share a lot in common with Class attributes.

**Reply by ActivistInvestor** [reply]
> > What does try, catch, and finally not have in common with these constructs
> 
> Parsing ambiguities.  As explicitly stated by [Bjarne Stroustrup](https://www.stroustrup.com/except89.pdf), even if the compiler could have rules which would make it unambiguous to parse, it would still be ambiguous and difficult to read.
> 
> > There seems to be no reason to allow these ambiguities even if there is a trivial and systematic way for a
> parser to chose one interpretation over another.

So Wirth, Hejlsberg, Tesler, et. al, had it all wrong?

**Reply by HaloFour** [reply]
> So Wirth, Hejlsberg, Tesler, et. al, had it all wrong?

I don't know what you're implying there.  Stroustrup designed C++ and laid out his reasoning for how the `try`/`catch` syntax was designed, particularly around why it explicitly needs blocks where most control statements suffice with embedded statements.  It doesn't seem Hejlsberg felt that needed to be readdressed when designing C#.

**Reply by ActivistInvestor** [reply]
In modern day Pascal, try/except/finally are implicitly block delimiters. So, no need for begin...end; to delimit a block of code inside of a try...except...finally, whether it is a single statement or multiple statements it is the same semantics.

**Reply by CyrusNajmabadi** [reply]
I don't really understand the pascal comparisons.  In pascal the `try` is required afaik.

**Reply by HaloFour** [reply]
Pascal also requires `end` at the end of a `try` block.

But that isn't particularly relevant to C#, and, as stated, Hejlsberg apparently did not feel that this was something worth reconsidering when designing C#.

**Reply by ActivistInvestor** [reply]
> > So Wirth, Hejlsberg, Tesler, et. al, had it all wrong?
> 
> It doesn't seem Hejlsberg felt that needed to be readdressed when designing C#.

That is what I am questioning. If a try block must always be followed by a pair of braces containing zero or more statements, which in-turn must be followed by either catch or finally, then what is the purpose of the braces?

It would seem to me that he simply followed the C++ convention.

IMO, In absence of a legitimate reason (not to suggest that there isn't one, just that I'm not aware of any), the Pascal convention is superior.

**Reply by CyrusNajmabadi** [reply]
> then what is the purpose of the braces?

To delineate the section of statements.

>  the Pascal convention is superior.

We did not think so when designing the language.  We preferred to consistently use braces, instead of having a disconnect where some blocks in a try were brace-less, and some were braced.

in other words, since `catch` and `finally` would already need braces to indicate where they terminated, leaving it out of `try would just be enormously inconsistent.

**Reply by CyrusNajmabadi** [reply]
> the Pascal convention is superior.

Pascal favors explicit end elements for many of its constructs.  C# went a different direction.  If you'd like a langauge more pascal-like in that regard, you could use VB.net.   It is a non-goal for C# to syntactically operate in a similar fashion.  We've explicitly chosen a style we prefer, and we try to use that consistently versus mixing the two different approaches together.

**Reply by HaloFour** [reply]
There is a well-supported Object Pascal compiler for .NET called [Oxygene](https://www.remobjects.com/elements/Oxygene/language.aspx) if you feel that it has superior syntax.

**Reply by ActivistInvestor** [reply]
> Pascal also requires `end` at the end of a `try` block.
> 
> But that isn't particularly relevant to C#, and, as stated, Hejlsberg apparently did not feel that this was something worth reconsidering when designing C#.

Yes, but my comment was about requiring block delimiters between try...except or try... finally. Syntactically, except and finally are components of try, which makes sense to me at least.

**Reply by CyrusNajmabadi** [reply]
As mentioned, i don't see what pascal has to do with this discussion.  Its stylistic design choices are very different (and, frankly, in opposition) to the lang style choices we made for c#.

**Reply by ActivistInvestor** [reply]
> > then what is the purpose of the braces?
> 
> To delineate the section of statements.
> 
> >  the Pascal convention is superior.
> 
> We did not think so when designing the language.  We preferred to consistently use braces, instead of having a disconnect where some blocks in a try were brace-less, and some were braced.
> 
> in other words, since `catch` and `finally` would already need braces to indicate where they terminated, leaving it out of `try would just be enormously inconsistent.  
> 

The Proposal originally was for not requiring braces at the outermost-level of a containing block. IOW, an enclosing block would delimit catch and finally.

**Reply by CyrusNajmabadi** [reply]
Right.  I've given a reason for why I don't think that's a good idea.  The symmetry here of all the sections is desirable.  As is making `try` explicit at the start.

**Reply by ActivistInvestor** [reply]
In that case I think that we will just agree that we disagree on that. 

Your points remind me of the many debates and discussions on LISP, and specifically how much opposition there was to the 'super-parentheses' (a "]" that closes any number of open parens) . Most of the arguments against it were founded on its lack of 'symmetry'. But that was a long time ago.

**Reply by CyrusNajmabadi** [reply]
> Your points remind me of the many debates and discussions on LISP, and specifically how much opposition there was to the 'super-parentheses

Right.  Style is *very* important to a language.  Attempting to add forms to a language that stylistically clash is not likely to be successful, or well received.

**Reply by ActivistInvestor** [reply]
And I think that's where we disagree.  Style is dictated to a great extent by language syntax and semantics.- That's a chicken/egg thing.

**Reply by CyrusNajmabadi** [reply]
When I say "style", I literally mean syntax.

**Comment by Jeltz191** [comment]
We could adjust your idea to leverage the using keyword to ensure that catches apply just to the scope of the using code?:
using catch(Exception c) =>UserCatchMethod(c)
using finally=>UserFinalCode();

**Reply by ActivistInvestor** [reply]
I would find this a good alternative, except that 'using' might cause some confusion with its current definition. The original though was that try/catch/finally are implicitly delimiting the code blocks that appear between them. I think that could only work at the outermost level of a body of code.

**Comment by theunrepentantgeek** [comment]
> The try, catch and finally blocks are implicitly-delimited, requiring no braces.

Or, put another way ...

The try, catch and finally blocks are implicitly-delimited, requiring any developer reading the code to stop and puzzle through to work out the scopes.

Explicit braces are straightforward to understand and brook no ambiguity about the actual scope. This is a good thing, not a problem that needs to be solved.

Trivial examples aside, production code is hard enough to understand as it is - I don't understand why it's a good idea to make it harder to read.

**Reply by ActivistInvestor** [reply]
You're failing to consider the circumstances under which The Proposal I made would be used. Typically it would involve a single statement in the try block and a single statement in the finally or catch block.

**Reply by CyrusNajmabadi** [reply]
@ActivistInvestor please keep things civil

**Reply by theunrepentantgeek** [reply]
>Typically it would involve a single statement in the try block and a single statement in the finally or catch block.

That's how you're thinking of it, sure. 
But is it realistic to assume that any developer using this feature would share the exact coding style as yourself? 

> Anyone that is incapable of figuring that out with an implicit block ... 

I'm sure almost any developer would be able to work it out if they gave it their entire attention.

However, I don't think that is at all likely. A developer coming in to read the method doesn't arrive with the thought _"I've got to work out how the implicit scopes work"_, they arrive with _"Why on earth does this code do THAT odd things that's caused an outage?"_ - or some similar high level goal. 

They shouldn't have to down tools and spend time puzzling through the code, just because someone decided a single level of indentation was a problem.
